agentcfg/model.go

```diff
@@ -18,32 +18,51 @@
 package agentcfg
 
 import (
+	"crypto/md5"
 	"encoding/json"
 	"fmt"
-	"strings"
-)
-
-const (
-	// ServiceName keyword
-	ServiceName = "service.name"
-	// ServiceEnv keyword
-	ServiceEnv = "service.environment"
 )
 
 // Doc represents an elasticsearch document
 type Doc struct {
-	ID     string `json:"_id"`
-	Source Source `json:"_source"`
-}
-
-// Source represents the elasticsearch _source field of a document
-type Source struct {
-	Settings Settings `json:"settings"`
+	Settings Settings
+	ID       string
 }
 
 // Settings hold agent configuration
 type Settings map[string]string
 
+// NewDoc unmarshals given byte slice into t Doc instance
+func NewDoc(in []byte) (*Doc, error) {
+	var doc Doc
+	if len(in) == 0 {
+		return &doc, nil
+	}
+	var tmp tmpDoc
+	if err := json.Unmarshal(in, &tmp); err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(tmp.Source.Settings, &doc.Settings); err != nil {
+		return nil, err
+	}
+	doc.ID = tmp.hash()
+	return &doc, nil
+}
+
+type tmpDoc struct {
+	ID     string `json:"_id"`
+	Source struct {
+		Settings json.RawMessage `json:"settings"`
+	} `json:"_source"`
+}
+
+func (t *tmpDoc) hash() string {
+	h := md5.New()
+	h.Write([]byte(t.ID))
+	h.Write([]byte(t.Source.Settings))
```

Comment:
are the settings guaranteed to be in the same order every time they are queried?

Reply:
Good question, I expected everything within `_source` field to be the unchanged stored json document and therefore to be returned in the same order for the same stored version. The server is not messing with it before the hash sum is calculated. @ogupte can you confirm that the `_source.settings` is always returned in the same order?
Since there is no version field and the timestamp doesn't get updated on editing, I don't see any other solution atm.

Reply:
Another solution would be to unmarshal the setttings, and loop over the keys alphabetically, adding the key and value to the hash along the way.  That would be and issue for nested keys but I"m not sure if we're planning on those at any point?